File: typescript/src/main/vsdk/toolkit/tonemap/FerwerdaToneMap.ts
import { Cie } from "../common/Cie";
import { ColorRgb } from "../common/ColorRgb";
import { ToneMap } from "./ToneMap";
import { ToneMappingContext } from "./ToneMappingContext";

export class FerwerdaToneMap extends ToneMap {
  private sf: ColorRgb;
  private msf: number;
  private pmComp: number;
  private pmDisplay: number;
  private smComp: number;
  private smDisplay: number;
  private lda: number;

  private static photopicOperator(logLa: number): number {
    let r: number;
    if (logLa <= -2.6) {
      r = -0.72;
    }
    else if (logLa >= 1.9) {
      r = logLa - 1.255;
    }
    else {
      r = globalThis.Math.pow(0.249 * logLa + 0.65, 2.7) - 0.72;
    }

    return globalThis.Math.pow(10.0, r);
  }

  private static scotopicOperator(logLa: number): number {
    let r: number;
    if (logLa <= -3.94) {
      r = -2.86;
    }
    else if (logLa >= -1.44) {
      r = logLa - 0.395;
    }
    else {
      r = globalThis.Math.pow(0.405 * logLa + 1.6, 2.18) - 2.86;
    }

    return globalThis.Math.pow(10.0, r);
  }

  private static mesopicScaleFactor(logLwa: number): number {
    if (logLwa < -2.5) {
      return 1.0;
    }
    if (logLwa > 0.8) {
      return 0.0;
    }
    return (0.8 - logLwa) / 3.3;
  }

  public constructor() {
    super();
    this.sf = new ColorRgb(0.062, 0.608, 0.330);
    this.msf = 0.0;
    this.pmComp = 0.0;
    this.pmDisplay = 0.0;
    this.smComp = 0.0;
    this.smDisplay = 0.0;
    this.lda = 0.0;
  }

  public override init(toneMapOptions: ToneMappingContext): void {
    const realWorldAdaptionLuminance = toneMapOptions.realWorldAdaptionLuminance;
    const maximumDisplayLuminance = toneMapOptions.maximumDisplayLuminance;
    this.lda = maximumDisplayLuminance / 2.0;

    this.msf = FerwerdaToneMap.mesopicScaleFactor(globalThis.Math.log10(realWorldAdaptionLuminance));
    this.smComp = FerwerdaToneMap.scotopicOperator(globalThis.Math.log10(this.lda)) /
      FerwerdaToneMap.scotopicOperator(globalThis.Math.log10(realWorldAdaptionLuminance));
    this.pmComp = FerwerdaToneMap.photopicOperator(globalThis.Math.log10(this.lda)) /
      FerwerdaToneMap.photopicOperator(globalThis.Math.log10(realWorldAdaptionLuminance));
    this.smDisplay = this.smComp / maximumDisplayLuminance;
    this.pmDisplay = this.pmComp / maximumDisplayLuminance;
  }

  public override scaleForComputations(radiance: ColorRgb): ColorRgb {
    const p = new ColorRgb();
    let sl: number;

    const eff = Cie.getLuminousEfficacy();
    radiance.scale(eff);

    p.set(radiance.r, radiance.g, radiance.b);
    sl = this.smComp * this.msf * (p.r * this.sf.r + p.g * this.sf.g + p.b * this.sf.b);

    radiance.scale(this.pmComp);

    if (sl > 0.0) {
      radiance.addConstant(radiance, sl);
    }

    return radiance;
  }

  public override scaleForDisplay(radiance: ColorRgb): ColorRgb {
    const p = new ColorRgb();
    let sl: number;

    const eff = Cie.getLuminousEfficacy();
    radiance.scale(eff);

    radiance.set(p.r, p.g, p.b);
    sl = this.smDisplay * this.msf * (p.r * this.sf.r + p.g * this.sf.g + p.b * this.sf.b);

    radiance.scale(this.pmDisplay);

    if (sl > 0.0) {
      radiance.addConstant(radiance, sl);
    }

    return radiance;
  }
}

File: cpp/src/tonemap/FerwerdaToneMap.cpp
#include "common/color/Cie.h"
#include "tonemap/FerwerdaToneMap.h"

/**
References:
[FERW1996] J.A. Ferwerda, S.N. Pattanaik, P. Shirley, D. Greenberg. A Model of
Visual Adaptation for Realistic Image Synthesis, SIGGRAPH 1996,
pp. 249-258.
*/

FerwerdaToneMap::FerwerdaToneMap():
    sf(0.062f, 0.608f, 0.330f),
    msf(),
    pmComp(),
    pmDisplay(),
    smComp(),
    smDisplay(),
    lda()
{
}

FerwerdaToneMap::~FerwerdaToneMap() {
}

void
FerwerdaToneMap::init(const ToneMappingContext &toneMapOptions) {
    const float realWorldAdaptionLuminance = toneMapOptions.realWorldAdaptionLuminance;
    const float maximumDisplayLuminance = toneMapOptions.maximumDisplayLuminance;
    lda = maximumDisplayLuminance / 2.0f;

    // Equations [FERW1996](4) and [FERW1996](5): t_p(L_a), t_s(L_a)
    msf = FerwerdaToneMap::mesopicScaleFactor(java::Math::log10(realWorldAdaptionLuminance));
    // Equation [FERW1996](3): m = t(L_da) / t(L_wa) using scotopic t_s from Equation (5)
    smComp = FerwerdaToneMap::scotopicOperator(java::Math::log10(lda)) /
             FerwerdaToneMap::scotopicOperator(java::Math::log10(realWorldAdaptionLuminance));
    // Equation [FERW1996](3): m = t(L_da) / t(L_wa) using photopic t_p from Equation (4)
    pmComp = FerwerdaToneMap::photopicOperator(java::Math::log10(lda)) /
             FerwerdaToneMap::photopicOperator(java::Math::log10(realWorldAdaptionLuminance));
    smDisplay = smComp / maximumDisplayLuminance;
    pmDisplay = pmComp / maximumDisplayLuminance;
}

ColorRgb
FerwerdaToneMap::scaleForComputations(ColorRgb radiance) const {
    ColorRgb p{};
    float sl;

    // Convert to photometric values
    const float eff = Cie::getLuminousEfficacy();
    radiance.scale(eff);

    // Compute the scotopic grayscale shift
    p.set(radiance.r, radiance.g, radiance.b);
    // Equation [FERW1996](6): L_d = L_dp + k(L_a) * L_ds
    sl = smComp * msf * (p.r * sf.r + p.g * sf.g + p.b * sf.b);

    // Scale the photopic luminance
    radiance.scale(pmComp);

    // Eventually, offset by the scotopic luminance
    if ( sl > 0.0 ) {
        radiance.addConstant(radiance, sl);
    }

    return radiance;
}

ColorRgb
FerwerdaToneMap::scaleForDisplay(ColorRgb radiance) const {
    ColorRgb p{};
    float sl;

    // Convert to photometric values
    const float eff = Cie::getLuminousEfficacy();
    radiance.scale(eff);

    // Compute the scotopic grayscale shift
    radiance.set(p.r, p.g, p.b);
    // Equation [FERW1996](6): L_d = L_dp + k(L_a) * L_ds
    sl = smDisplay * msf * (p.r * sf.r + p.g * sf.g + p.b * sf.b);

    // Scale the photopic luminance
    radiance.scale(pmDisplay);

    // Eventually, offset by the scotopic luminance
    if ( sl > 0.0 ) {
        radiance.addConstant(radiance, sl);
    }

    return radiance;
}

float
FerwerdaToneMap::photopicOperator(float logLa) {
    // Equation [FERW1996](4): piecewise approximation of log(t_p(L_a))
    float r;
    if ( logLa <= -2.6 ) {
        r = -0.72f;
    } else if ( logLa >= 1.9 ) {
        r = logLa - 1.255f;
    } else {
        r = java::Math::pow(0.249f * logLa + 0.65f, 2.7f) - 0.72f;
    }

    return java::Math::pow(10.0f, r);
}

float
FerwerdaToneMap::scotopicOperator(float logLa) {
    // Equation [FERW1996](5): piecewise approximation of log(t_s(L_a))
    float r;
    if ( logLa <= -3.94 ) {
        r = -2.86f;
    } else if ( logLa >= -1.44 ) {
            r = logLa - 0.395f;
        } else {
            r = java::Math::pow(0.405f * logLa + 1.6f, 2.18f) - 2.86f;
        }

    return java::Math::pow(10.0f, r);
}

float
FerwerdaToneMap::mesopicScaleFactor(float logLwa) {
    if ( logLwa < -2.5 ) {
        return 1.0;
    } else if ( logLwa > 0.8 ) {
            return 0.0;
        } else {
            return (0.8f - logLwa) / 3.3f;
        }
}
